service/labelstore/service.go

```diff
@@ -196,26 +198,23 @@ func (s *service) GetLocalRefID(componentID string, globalRefID uint64) uint64 {
 	return local
 }
 
-// AddStaleMarker adds a stale marker
-func (s *service) AddStaleMarker(globalRefID uint64, l labels.Labels) {
+func (s *service) TrackStaleness(ids []StalenessTracker) {
 	s.mut.Lock()
 	defer s.mut.Unlock()
 
-	s.staleGlobals[globalRefID] = &staleMarker{
-		lastMarkedStale: time.Now(),
-		labelHash:       l.Hash(),
-		globalID:        globalRefID,
+	for _, id := range ids {
+		if value.IsStaleNaN(id.Value) {
+			s.staleGlobals[id.GlobalRefID] = &staleMarker{
+				lastMarkedStale: time.Now(),
+				labelHash:       id.Labels.Hash(),
```

Comment:
Let's make sure we don't call `id.Labels.Hash()` inside the critical section. Something like this:
```
func (s *service) TrackStaleness(ids []StalenessTracker) {
	var (
		toAdd    = make([]*staleMarker, 0)
		toRemove = make([]uint64, 0)
		now      = time.Now()
	)

	for _, id := range ids {
		if value.IsStaleNaN(id.Value) {
			toAdd = append(toAdd, &staleMarker{
				globalID:        id.GlobalRefID,
				lastMarkedStale: now,
				labelHash:       id.Labels.Hash(),
			})
		} else {
			toRemove = append(toRemove, id.GlobalRefID)
		}
	}

	s.mut.Lock()
	defer s.mut.Unlock()

	for _, marker := range toAdd {
		s.staleGlobals[marker.globalID] = marker
	}
	for _, id := range toRemove {
		delete(s.staleGlobals, id)
	}
}
```

Reply:
Sure that makes sense.